src/hotspot/share/memory/metaspace.cpp

```diff
@@ -627,22 +627,35 @@ ReservedSpace Metaspace::reserve_address_space_for_compressed_classes(size_t siz
     log_debug(metaspace, map)("Trying between " UINT64_FORMAT_X " and " UINT64_FORMAT_X
                               " with " SIZE_FORMAT_X " alignment", min, max, alignment);
     result = os::attempt_reserve_memory_between((char*)min, (char*)max, size, alignment, randomize);
+
+#ifdef AARCH64
+    if (result == nullptr) {
```

Comment:
Why not do the optimized (for AArch64) stuff before trying the immediate + shift?

Reply:
> Why not do the optimized (for AArch64) stuff before trying the immediate + shift?

For AArch64, the optimum is zero base, no shift, then immediate base, no shift, then shift, and last of all immediate + shift.

Reply:
> > > Why not do the optimized (for AArch64) stuff before trying the immediate + shift?
> > 
> > For AArch64, the optimum is zero base, no shift, then immediate base, no shift, then shift, and last of all immediate + shift.
> 
> Wait, immediate + shift=0 before no immediate=0 + shift? Why? Looking at https://dougallj.github.io/applecpu/firestorm-int.html , both movk and lsl need the same number of cycles. Why would movk be better than lsl?

I agree, good point. Those are equal on modern AArch64 CPUs. LSL (immediate) is an alias for UBFM, and I had a vague memory that was slower, but I don't think that's so now, if ever it was.

> The reason I ask is that this would mean switching up the order of things I do in Metaspace::reserve_address_space_for_compressed_classes. That function attempts to do the best for all platforms, and all other platforms prefer zero immediate + shift. It would mean even more ifdefs than we have before.

I see.

> (the best way would be to move this function into CPU specific files. We would have some duplication, but at least every CPU can then decide how to layout memory, and the comments would not have to be so darn vague. 

Ah, it's not just me, then.

I think you at least have to say what you think might fail, rather than we can say no more. Something like "If the address space is already highly allocated we might fail." 

> But last time I attempted that I ran into objections from reviewers).

Shame. This is very CPU specific.

Reply:
>> (the best way would be to move this function into CPU specific files. We would have some duplication, but at least every CPU can then decide how to layout memory, and the comments would not have to be so darn vague.

> Ah, it's not just me, then.

No, I really dislike this too. Forcing something that is inherently CPU specific in a CPU agnostic form, and then having to explain it every time.

> Shame. This is very CPU specific.

I may retry making this CPU specific, once I have some cycles left.